File: metrics.py
"""Derive the specs.md §20 metrics from ``events.csv`` alone (T6.2).

CC-06 requires that every metric in specs.md §20 be computable from the event
log, with nothing existing only in memory and printed. That requirement is easy
to *state* and easy to quietly violate, so this module is the check: it takes
event rows and nothing else, and produces the metrics. The endpoints compute the
same numbers as they run, and a test asserts the two agree — if a metric were
only ever computed in memory, the derivation here would have nothing to read.

It is deliberately standard library only, like the protocol itself. The pandas
aggregation of T9.1 is a different job: it spans runs, averages trials and draws
graphs. This spans one log and answers "what does this run say about itself".

Reading rule (RP-07): logs are opened read-only and never modified.

WHAT NEEDS WHICH ENDPOINT
=========================
The two endpoints see different things, and no metric is invented from the
wrong side:

* the **sender's** log carries SEND / RETX and their byte counts, so
  retransmission count, retransmission overhead and the RTT samples come from
  there, as do mode residence and switch count;
* the **receiver's** log carries DELIVER, so delivered application bytes — and
  therefore goodput — come from there.

Passing both logs fills in everything. Passing one fills in what that endpoint
can honestly answer and leaves the rest as None, rather than guessing.

TIMESTAMPS
==========
Each log's timestamps are seconds from *that endpoint's* start, on the
``time.perf_counter`` clock, so the two are not on a common origin. Every figure
below is computed within a single log, and the two are never subtracted from one
another. To interleave them for reading, align on the START row the two share.
"""

from __future__ import annotations

import csv
import statistics
from pathlib import Path
from typing import Any, Iterable, Sequence

#: A SWITCH row with this in its reason is the ``fixed-hybrid`` control paying
#: the drain cost without changing mode (T5.6). It is a handshake, not a
#: transition, and must not be counted as one.
NOOP_SWITCH_MARKER = "FIXED_HYBRID_NOOP"

#: Modes a transfer can reside in. "saw" is the Phase 2 placeholder; it is
#: reported if it appears, but it is not a system under evaluation.
RESIDENCE_MODES = ("gbn", "sr", "saw")


def read_events(path: Path | str) -> list[dict[str, str]]:
    """Read one ``events.csv``. Accepts the file or the directory holding it."""
    path = Path(path)
    if path.is_dir():
        path = path / "events.csv"
    with open(path, newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def _number(row: dict[str, str], column: str) -> float | None:
    value = row.get(column, "")
    return float(value) if value not in ("", None) else None


def _rows(events: Iterable[dict[str, str]], name: str) -> list[dict[str, str]]:
    return [row for row in events if row["event"] == name]


def _endpoint_rows(events: Iterable[dict[str, str]], endpoint: str) -> list[dict[str, str]]:
    return [row for row in events if row.get("endpoint") == endpoint]


# ---------------------------------------------------------------------------
# Individual metrics
# ---------------------------------------------------------------------------


def retransmission_count(events: Sequence[dict[str, str]]) -> int:
    """specs.md §20: total DATA retransmission events — a count of RETX rows.

    SEND and RETX are distinct events precisely so this is a count and not a
    reconstruction (design.md §9, TO-05).
    """
    return len(_rows(events, "RETX"))


def retransmission_overhead(events: Sequence[dict[str, str]]) -> float | None:
    """Retransmitted bytes over total transmitted bytes.

    Both sums come from the ``bytes`` column of SEND and RETX, which is the
    datagram size as it went on the wire — headers included, because that is
    what "transmitted" means on a network.
    """
    sent = sum(_number(row, "bytes") or 0.0 for row in _rows(events, "SEND"))
    resent = sum(_number(row, "bytes") or 0.0 for row in _rows(events, "RETX"))
    total = sent + resent
    return (resent / total) if total else None


def delivered_bytes(events: Sequence[dict[str, str]]) -> int | None:
    """Application bytes written to the output file, from the receiver's DELIVER rows.

    The receiver's writer refuses a duplicate, and only a segment it actually
    wrote produces a DELIVER — so this cannot double count a retransmission
    (T2.3, SEQ-05).
    """
    rows = _rows(events, "DELIVER")
    if not rows:
        return None
    return int(sum(_number(row, "bytes") or 0.0 for row in rows))


def integrity_success(events: Sequence[dict[str, str]]) -> bool | None:
    """Whether source and received hashes matched, from the FIN_ACK verdict.

    A hash mismatch is a reported failure, never silence (CC-01), so its absence
    from the log is *not* read as success: a run with no verdict returns None.
    """
    verdicts = [row["reason"] for row in _rows(events, "FIN_ACK") if row["reason"]]
    if not verdicts:
        return None
    return any(reason.startswith("MATCH") for reason in verdicts)


def completion_time_s(events: Sequence[dict[str, str]]) -> float | None:
    """Time from the start of the log to verified completion.

    The end is the FIN_ACK that carried the verdict; a run that never got one
    ends at its last event, which is the honest reading of when it stopped.
    """
    if not events:
        return None
    verdicts = _rows(events, "FIN_ACK")
    last = verdicts[0] if verdicts else events[-1]
    return _number(last, "timestamp")


def goodput_bytes_per_s(events: Sequence[dict[str, str]],
                        elapsed_s: float | None = None) -> float | None:
    """Delivered application bytes per second (specs.md §20).

    A transfer whose integrity check failed has no meaningful throughput, so it
    is reported as None rather than averaged in later (design.md §10).
    """
    delivered = delivered_bytes(events)
    elapsed = elapsed_s if elapsed_s is not None else completion_time_s(events)
    if delivered is None or not elapsed:
        return None
    if integrity_success(events) is False:
        return None
    return delivered / elapsed


def switch_count(events: Sequence[dict[str, str]]) -> int:
    """GBN<->SR transitions: SWITCH rows, excluding the fixed-hybrid no-ops."""
    return len([row for row in _rows(events, "SWITCH")
                if NOOP_SWITCH_MARKER not in row["reason"]])


def handshake_count(events: Sequence[dict[str, str]]) -> int:
    """Completed MODE exchanges, whether or not they changed the mode."""
    return len(_rows(events, "SWITCH"))


def mode_residence(events: Sequence[dict[str, str]]) -> dict[str, float]:
    """Seconds spent in each mode, from the ``mode`` column over time.

    Every row carries the mode that was live when it was written, so residence
    is the time between the first and last row of each contiguous run of a mode
    — no separate bookkeeping, and nothing that exists only in memory.
    """
    timed = [(_number(row, "timestamp"), row["mode"]) for row in events if row["mode"]]
    timed = [(moment, mode) for moment, mode in timed if moment is not None]
    residence: dict[str, float] = {}
    for (start, mode), (end, _) in zip(timed, timed[1:]):
        residence[mode] = residence.get(mode, 0.0) + max(0.0, end - start)
    return residence


def rtt_samples_ms(events: Sequence[dict[str, str]]) -> list[float]:
    """Per-segment round trips, in milliseconds, from SEND and ACK rows.

    Karn's rule (design.md §5.4): a segment that was retransmitted has an
    ambiguous ACK and is excluded. Only an ACK naming the segment exactly is
    used, which under SR means every segment and under GBN means the subset
    whose cumulative ACK happens to name them — fewer samples, each still
    honest. That is why §20 calls these statistics optional.
    """
    sent_at: dict[str, float] = {}
    for row in _rows(events, "SEND"):
        moment = _number(row, "timestamp")
        if moment is not None:
            sent_at.setdefault(row["sequence"], moment)
    for row in _rows(events, "RETX"):
        sent_at.pop(row["sequence"], None)      # ambiguous; excluded

    samples = []
    seen: set[str] = set()
    for row in _rows(events, "ACK"):
        key = row["ack"]
        if key in sent_at and key not in seen:
            moment = _number(row, "timestamp")
            if moment is not None:
                samples.append((moment - sent_at[key]) * 1000.0)
                seen.add(key)
    return samples


def rtt_statistics(events: Sequence[dict[str, str]]) -> dict[str, float | None]:
    """Mean, median and tail RTT (specs.md §20, optional statistics)."""
    samples = rtt_samples_ms(events)
    if not samples:
        return {"rtt_samples": 0, "rtt_mean_ms": None, "rtt_median_ms": None,
                "rtt_p95_ms": None, "rtt_max_ms": None}
    ordered = sorted(samples)
    index = min(len(ordered) - 1, int(round(0.95 * (len(ordered) - 1))))
    return {
        "rtt_samples": len(samples),
        "rtt_mean_ms": statistics.fmean(samples),
        "rtt_median_ms": statistics.median(samples),
        "rtt_p95_ms": ordered[index],
        "rtt_max_ms": ordered[-1],
    }


def delivery_latency_s(events: Sequence[dict[str, str]]) -> dict[str, float | None]:
    """Latency "where defined" (specs.md §20): time to first and last delivery.

    Measured inside the receiver's log, from its first event to the DELIVER of
    the first and last segment. Per-segment sender-to-receiver latency is *not*
    reported: the two logs have different origins, and subtracting across them
    would produce a number that looks precise and means nothing.
    """
    delivered = _rows(events, "DELIVER")
    if not delivered:
        return {"first_delivery_s": None, "last_delivery_s": None}
    return {
        "first_delivery_s": _number(delivered[0], "timestamp"),
        "last_delivery_s": _number(delivered[-1], "timestamp"),
    }


# ---------------------------------------------------------------------------
# The whole of specs.md §20
# ---------------------------------------------------------------------------


def derive_metrics(events: Sequence[dict[str, str]]) -> dict[str, Any]:
    """Every specs.md §20 metric this log can answer, and None for the rest.

    Rows from both endpoints may be passed together; each metric is taken from
    the endpoint that actually observed it.
    """
    sender = _endpoint_rows(events, "sender") or list(events)
    receiver = _endpoint_rows(events, "receiver") or list(events)

    elapsed = completion_time_s(sender)
    residence = mode_residence(sender)
    metrics: dict[str, Any] = {
        "completion_time_s": elapsed,
        "goodput_bytes_per_s": goodput_bytes_per_s(receiver, elapsed),
        "delivered_bytes": delivered_bytes(receiver),
        "retransmission_count": retransmission_count(sender),
        "retransmission_overhead": retransmission_overhead(sender),
        "unique_data_packets": len(_rows(sender, "SEND")),
        "total_data_transmissions": len(_rows(sender, "SEND")) + len(_rows(sender, "RETX")),
        "bytes_transmitted": int(sum(
            (_number(row, "bytes") or 0.0)
            for row in _rows(sender, "SEND") + _rows(sender, "RETX"))),
        "integrity_success": integrity_success(events),
        "switch_count": switch_count(sender),
        "handshake_count": handshake_count(sender),
        "gbn_residence_s": residence.get("gbn", 0.0),
        "sr_residence_s": residence.get("sr", 0.0),
    }
    metrics.update(rtt_statistics(sender))
    metrics.update(delivery_latency_s(receiver))
    return metrics


def derive_from_run(run_directory: Path | str) -> dict[str, Any]:
    """Derive the metrics for one run from ``logs/<run_id>/``, both endpoints.

    Opens the logs read-only and leaves them exactly as they were (RP-07).
    """
    run_directory = Path(run_directory)
    events: list[dict[str, str]] = []
    for endpoint in ("sender", "receiver"):
        path = run_directory / endpoint / "events.csv"
        if path.exists():
            events.extend(read_events(path))
    if not events:
        raise FileNotFoundError(f"no events.csv under {run_directory}")
    return derive_metrics(events)

File: frontend/static/js/views/metrics.js
/* Metrics Dashboard (T11.9): every specs.md §20 metric, with units.
 *
 * The numbers here are the ones metrics.py derives from events.csv, computed on
 * the server by that module — not re-derived in the browser. That is what makes
 * "a completed run's numbers match what metrics.py derives for the same run"
 * true by construction rather than by coincidence.
 *
 * Where a panel does arithmetic of its own — a ratio for display, a percentage
 * of a total — it is labelled a presentation-only calculation, so no reader
 * mistakes it for a recorded measurement.
 */

import { api } from '../api.js';
import {
  el, clear, card, stat, badge, viewHeader, emptyState, errorState, integrityBadge, table,
  num, bytes, rate, seconds, percent, MISSING
} from '../ui.js';

let body = null;
let cachedKey = null;

const SPEC_20 = [
  ['Goodput', 'Successfully delivered application bytes / transfer time'],
  ['Completion time', 'Time from transfer start to verified completion'],
  ['Retransmission count', 'Total DATA retransmission events'],
  ['Retransmission overhead', 'Retransmitted bytes / total transmitted bytes'],
  ['Integrity success', 'Whether source and received hashes match'],
  ['Switch count', 'Total GBN↔SR transitions'],
  ['GBN residence', 'Time spent in GBN'],
  ['SR residence', 'Time spent in SR'],
  ['Latency', 'Measured delivery/transfer delay where defined'],
  ['Optional RTT statistics', 'Mean, median, tail RTT'],
];

function headline(derived) {
  return el('div', { class: 'grid cols-4' },
    stat('Goodput', rate(derived.goodput_bytes_per_s),
      { sub: 'delivered bytes / completion time' }),
    stat('Completion time', seconds(derived.completion_time_s),
      { sub: 'to the FIN_ACK that carried the verdict' }),
    stat('Retransmissions', num(derived.retransmission_count, 0),
      { sub: 'count of RETX rows' }),
    stat('Retransmission overhead', percent(derived.retransmission_overhead),
      { sub: 'resent bytes / all transmitted bytes' }));
}

function residenceBar(derived) {
  const gbn = Number(derived.gbn_residence_s) || 0;
  const sr = Number(derived.sr_residence_s) || 0;
  const total = gbn + sr;
  if (!total) return el('p', { class: 'note-inline' }, 'No mode residence was recorded.');
  return el('div', {},
    el('div', { class: 'progress', style: 'height:18px' },
      el('span', {
        style: `width:${((gbn / total) * 100).toFixed(2)}%;background:var(--gbn)`,
        title: `GBN ${gbn.toFixed(3)} s`,
      })),
    el('p', { class: 'note-inline' },
      `GBN ${seconds(gbn)} (${percent(gbn / total)}) · SR ${seconds(sr)} (${percent(sr / total)})`,
      el('span', { class: 'stat-sub' },
        ' — the fractions are a presentation-only calculation over the two recorded seconds.')));
}

function specTable(derived) {
  const rows = [
    ['Goodput', rate(derived.goodput_bytes_per_s), 'bytes/s', 'receiver DELIVER rows ÷ sender completion time'],
    ['Completion time', seconds(derived.completion_time_s), 's', 'sender log, to FIN_ACK'],
    ['Retransmission count', num(derived.retransmission_count, 0), 'events', 'RETX rows'],
    ['Retransmission overhead', percent(derived.retransmission_overhead), 'fraction', 'bytes column of RETX ÷ SEND+RETX'],
    ['Integrity success', integrityBadge(derived.integrity_success), '', 'FIN_ACK verdict'],
    ['Switch count', num(derived.switch_count, 0), 'transitions', 'SWITCH rows, excluding FIXED_HYBRID_NOOP'],
    ['MODE handshakes', num(derived.handshake_count, 0), 'exchanges', 'all SWITCH rows'],
    ['GBN residence', seconds(derived.gbn_residence_s), 's', 'mode column over time'],
    ['SR residence', seconds(derived.sr_residence_s), 's', 'mode column over time'],
    ['Delivered bytes', bytes(derived.delivered_bytes), 'B', 'receiver DELIVER rows'],
    ['Bytes transmitted', bytes(derived.bytes_transmitted), 'B', 'sender SEND + RETX rows'],
    ['Unique DATA packets', num(derived.unique_data_packets, 0), 'packets', 'SEND rows'],
    ['Total DATA transmissions', num(derived.total_data_transmissions, 0), 'packets', 'SEND + RETX rows'],
    ['First delivery', seconds(derived.first_delivery_s), 's', 'receiver log, first DELIVER'],
    ['Last delivery', seconds(derived.last_delivery_s), 's', 'receiver log, last DELIVER'],
    ['RTT samples', num(derived.rtt_samples, 0), 'samples', 'unambiguous SEND→ACK pairs (Karn)'],
    ['RTT mean', num(derived.rtt_mean_ms, 2), 'ms', 'of those samples'],
    ['RTT median', num(derived.rtt_median_ms, 2), 'ms', 'of those samples'],
    ['RTT p95', num(derived.rtt_p95_ms, 2), 'ms', 'of those samples'],
    ['RTT max', num(derived.rtt_max_ms, 2), 'ms', 'of those samples'],
  ];

  return table([
    { label: 'Metric', render: (row) => row[0] },
    { label: 'Value', numeric: true, render: (row) => row[1] },
    { label: 'Unit', render: (row) => row[2] || MISSING },
    { label: 'Derived from', wrap: true, render: (row) => row[3] },
  ], rows, { maxHeight: '58vh' });
}

function agreementCard(detail) {
  /* The endpoints computed their own counters while they ran, and metrics.py
   * derives the same figures from the log afterwards. T6.2 asserts the two
   * agree; showing them side by side makes that checkable on the screen too. */
  const derived = detail.derived_metrics || {};
  const own = (detail.sender_summary || {}).metrics || {};
  const pairs = [
    ['completion_time_s', 'Completion time', (v) => seconds(v)],
    ['retransmission_count', 'Retransmissions', (v) => num(v, 0)],
    ['switch_count', 'Switches', (v) => num(v, 0)],
    ['goodput_bytes_per_s', 'Goodput', (v) => rate(v)],
  ].filter(([key]) => own[key] !== undefined);

  if (!pairs.length) {
    return card('Derived vs the endpoint’s own counters', {},
      emptyState('No sender summary for this run',
        'Only the derivation from events.csv is available, which is the authoritative one — '
        + 'CC-06 requires every §20 metric to be computable from the log alone.'));
  }

  return card('Derived vs the endpoint’s own counters', { note: 'T6.2 asserts these agree' },
    table([
      { label: 'Metric', render: (row) => row.label },
      { label: 'Derived from events.csv', numeric: true, render: (row) => row.derived },
      { label: 'Recorded in summary.json', numeric: true, render: (row) => row.own },
      { label: '', render: (row) => (row.agrees ? badge('agree', 'ok') : badge('differ', 'warn')) },
    ], pairs.map(([key, label, format]) => {
      const a = Number(derived[key]);
      const b = Number(own[key]);
      const agrees = !Number.isFinite(a) || !Number.isFinite(b)
        ? a === b : Math.abs(a - b) <= Math.max(1e-6, Math.abs(b) * 0.02);
      return { label, derived: format(derived[key]), own: format(own[key]), agrees };
    })),
    el('p', { class: 'note-inline' },
      'The left column is what <code>metrics.py</code> reads out of the event log; the right is '
      + 'what the endpoint counted as it ran. Comparison is to within 2%, because the two stop '
      + 'their clocks a fraction of a second apart.'));
}

async function renderBody(store) {
  const fragment = el('div', { class: 'grid' });
  if (!store.selectedKey) {
    fragment.append(emptyState('No run selected', 'Pick a run from the header, or start one.'));
    return fragment;
  }
  let detail;
  try {
    detail = await api.run(store.selectedKey);
  } catch (error) {
    fragment.append(errorState(error, { title: 'The metrics could not be read' }));
    return fragment;
  }
  cachedKey = store.selectedKey;
  const derived = detail.derived_metrics || {};

  if (!Object.keys(derived).length) {
    fragment.append(emptyState('This run has no event rows to derive from',
      'Every §20 metric is computed from events.csv alone (CC-06), so an empty log yields '
      + 'no metrics rather than zeros.'));
    return fragment;
  }

  fragment.append(card('Headline', {
    note: `run ${detail.run_id}`,
  }, headline(derived),
  el('div', { class: 'grid cols-4', style: 'margin-top:16px' },
    stat('Integrity', integrityBadge(derived.integrity_success), { sub: 'source vs received hash' }),
    stat('Switches', num(derived.switch_count, 0),
      { sub: `${num(derived.handshake_count, 0)} handshakes` }),
    stat('Delivered', bytes(derived.delivered_bytes), { sub: 'application bytes written' }),
    stat('Transmitted', bytes(derived.bytes_transmitted), { sub: 'datagram bytes on the wire' })),
  el('div', { style: 'margin-top:16px' },
    el('h4', {}, 'Mode residence'), residenceBar(derived)),
  derived.integrity_success === false
    ? el('p', { class: 'card-caveat', style: 'border-color:var(--bad)' },
      '<strong>Integrity failed for this run.</strong> Goodput is reported as unavailable rather '
      + 'than computed: a transfer that did not deliver the file has no meaningful throughput, '
      + 'and averaging one in later would flatter the result.')
    : null));

  fragment.append(el('div', { class: 'grid cols-2' },
    card('Every specs.md §20 metric', { note: 'derived by metrics.py from events.csv alone' },
      specTable(derived)),
    el('div', { class: 'grid' },
      agreementCard(detail),
      card('What §20 asks for', { note: 'the specification, verbatim' },
        table([
          { label: 'Metric', render: (row) => row[0] },
          { label: 'Definition', wrap: true, render: (row) => row[1] },
        ], SPEC_20)))));
  return fragment;
}

export async function render({ store }) {
  const root = el('div');
  root.append(viewHeader('Metrics Dashboard',
    'Every metric in specs.md §20, with units, for the selected run. These are the values '
    + '<code>metrics.py</code> derives from <code>events.csv</code> — the same derivation T6.2 '
    + 'asserts against the endpoints’ own counters.'));
  body = el('div', {}, await renderBody(store));
  root.append(body);
  return root;
}

let refreshing = false;

export async function onTick({ store }) {
  if (!body || refreshing) return;
  if (!store.transfer.active && store.selectedKey === cachedKey) return;
  refreshing = true;
  try { clear(body).append(await renderBody(store)); } finally { refreshing = false; }
}
